### PR TITLE
Do not update WindowTitle at all in prompt func if setting is $null

### LINKED ISSUE #594: WindowTitle set by user overwritten by module
### System Details

- posh-git version/path: 1.0.0 beta2 ~\Documents\WindowsPowerShell\Modules\posh-git\1.0.0
- PowerShell version: 5.1.17134.112
- git version 2.18.0.windows.1
- OS: Microsoft Windows NT 10.0.17134.0

### Issue Description

I have the following in my $profile:

    Import-Module posh-git
    $GitPromptSettings.WindowTitle = $null

If I start a script that changes `$host.UI.RawUI.WindowTitle` directly then this still gets overwritten with `$PreviousWindowTitle` that was initialised with the text from when the PowerShell window was started. I would like my script to be able to change the title and have the same effect regardless of whether posh-git is loaded.

I know there has been some discussion about this before in previous issues - but could it be considered that a setting could be added to `$GitPromptSettings` to disable any update of `$host.UI.RawUI.WindowTitle`?

Maybe this could also be added as a parameter when the module starts so that the WindowTitle never gets updated or changed by the module (and hence would not need to be reverted to $PreviousWindowTitle).

### PR DESCRIPTION
**Pull request by rkeithhill**
We used to restore the previousWindowTitle but this had the effect of
always updating the WindowTitle with a fixed string - whatever we
captured in $PreviousWindowTitle.

However, we still restore the original window title when the module is
removed. But we only do this if the module is configured to supply the
window title.

Rename variable to make its intent clearer.

Fixes #594

### CODE REVIEW

File: src/posh-git.psm1 [hunk 1]
```diff
@@ -18,16 +18,17 @@ if (!$Env:HOME) { $Env:HOME = "$Env:USERPROFILE" }
 $IsAdmin = Test-Administrator
 
 # Probe $Host.UI.RawUI.WindowTitle to see if it can be set without errors
-$WindowTitleSupported = $false
+$HostSupportsSettingWindowTitle = $false
 try {
-    $global:PreviousWindowTitle = $Host.UI.RawUI.WindowTitle
-    $newTitle = "${global:PreviousWindowTitle} "
+    $global:PoshGitOrigWindowTitle = $Host.UI.RawUI.WindowTitle
```
Comment (dahlbyk) [inline src/posh-git.psm1]: Does this actually need to be global? Could it be script-scoped alongside `$HostSupportsSettingWindowTitle`, with a corresponding `Reset-WindowTitle` to use in `OnRemove`?

File: src/posh-git.psm1 [hunk 2]
```diff
@@ -92,34 +89,22 @@ $GitPromptScriptBlock = {
...
 
-    # Update the host's WindowTitle is host supports it and user has not disabled $GitPromptSettings.WindowTitle
+    # Update the host's WindowTitle if host supports it and user has not disabled $GitPromptSettings.WindowTitle
     # This has to be *after* the call to Write-VcsStatus, which populates $global:GitStatus
-    if ($WindowTitleSupported) {
-        $windowTitle = $settings.WindowTitle
-        if (!$windowTitle) {
-            if ($global:PreviousWindowTitle) {
-                $Host.UI.RawUI.WindowTitle = $global:PreviousWindowTitle
+    if ($HostSupportsSettingWindowTitle -and $settings.WindowTitle) {
```
Comment (dahlbyk) [inline src/posh-git.psm1]: While we're at it, should this whole block move into a `Set-WindowTitle` function?

File: src/posh-git.psm1 [hunk 3]
```diff
@@ -18,16 +18,17 @@ if (!$Env:HOME) { $Env:HOME = "$Env:USERPROFILE" }
 $IsAdmin = Test-Administrator
 
 # Probe $Host.UI.RawUI.WindowTitle to see if it can be set without errors
-$WindowTitleSupported = $false
+$HostSupportsSettingWindowTitle = $false
```
[the remark this answers is not in the document]
Comment (rkeithhill) [reply src/posh-git.psm1]: But when `$settings.WindowTitle` isn't falsey, we would be doing this "writable" check during every prompt eval - effectively setting the title three times.  

I think it is better to do this check once during module load but I'm fine with creating a function in Utils.ps1 so we'd have:

```posh
$IsAdmin = Test-Administrator
$IsWindowTitleWritable = Test-WindowTitleIsWritable
```

File: src/posh-git.psm1 (hunk 1 above)
Comment (rkeithhill) [reply src/posh-git.psm1]: Yeah, it could be script scoped.

File: src/posh-git.psm1 (hunk 2 above)
Comment (rkeithhill) [reply src/posh-git.psm1]: Sure, but we'd need to export it from the module to make it callable from the prompt func.  How about `Reset-WindowTitle` - does it need to be exported or would we be the only caller in the OnRemove event handler.

**Comment by dahlbyk** [comment]
> > While we're at it, should this whole block move into a `Set-WindowTitle` function?
>
> Sure, but we'd need to export it from the module to make it callable from the prompt func.

I was pretty sure this wasn't the case, and my experiment to find out may have gone a bit far. Anyway, the new commits here seem to work as expected and the module is much cleaner.

image

File: src/WindowTitle.ps1 [hunk 4]
```diff
@@ -0,0 +1,60 @@
+$HostSupportsSettingWindowTitle = $null
+$OriginalWindowTitle = $null
+
+function Test-WindowTitleIsWriteable {
+    if ($null -eq $HostSupportsSettingWindowTitle) {
```
Comment (rkeithhill) [inline src/WindowTitle.ps1]: 👍

**Comment by rkeithhill** [comment]
Looks good!  My daughter just got back from her choir trip to Europe so didn't have time to tackle this last night.  Thanks!!  Go ahead and merge whenever you're ready.

I have to say I'm a bit surprised the scriptblock we assign to the global func prompt can see that non-exported command.  I was assuming the scriptblock closed over the variables it used.  TIL that scriptblock seems to have access to module scope.  :-)

**Pull request merged**